notebooks/byod.py

```diff
@@ -72,134 +98,136 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
-            del keyed_rows[key][key_column]
-        return keyed_rows
-
-    def keyed_row_columns(keyed_rows: Dict[str, Any]) -> Set[str]:
-        if keyed_rows:
-            random_key = set(keyed_rows.keys()).pop()
-            return set(keyed_rows[random_key].keys())
+    def parse_cram_crai(filename: str):
+        if filename.endswith(".cram.crai"):
+            # double extension in the crai file, ie "foo.cram.crai"
+            sample = filename[:-10]
         else:
-            return set()
-
-    def join_keyed_rows(keyed_rows_a: Dict[str, Any], keyed_rows_b: Dict[str, Any]) -> Dict[str, Any]:
-        a_columns = keyed_row_columns(keyed_rows_a)
-        b_columns = keyed_row_columns(keyed_rows_b)
-        assert not a_columns.intersection(b_columns), "Keyed rows to join may not share columns"
-        common_keys = set(keyed_rows_a.keys()).union(set(keyed_rows_b.keys()))
-        return {k: dict(**keyed_rows_a.get(k, {c: BLANK_CELL_VALUE for c in a_columns}),
-                        **keyed_rows_b.get(k, {c: BLANK_CELL_VALUE for c in b_columns}))
-                for k in common_keys}
-
-    def join_data_tables(new_table: str, tables_to_join: list, join_column: str):
-        keyed_rows = get_keyed_rows(tables_to_join[0], join_column)
-        for table_name in tables_to_join[1:]:
-            keyed_rows = join_keyed_rows(keyed_rows, get_keyed_rows(table_name, join_column))
-        upload_rows(new_table, [{join_column: k, **row} for k, row in keyed_rows.items()])
+            sample = filename[:-5]
+        ext = filename[-4:]
+        return sample, ext
 
     def create_cram_crai_table(table: str, listing: Iterable[str]):
-        crams = dict()
-        crais = dict()
+        matched_files = defaultdict(dict)  # type: dict
+        allCrams = []
+        allCrais = []
         for key in listing:
             _, filename = key.rsplit("/", 1)
-
-            parts = filename.split(".")
-            if 3 == len(parts):  # foo.cram.crai branch
-                sample, _, ext = parts
-            elif 2 == len(parts):  # "foo.cram" or "foo.crai" branch
-                sample, ext = parts
-            else:
-                raise ValueError(f"Unable to parse '{filename}'")
-
-            if "cram" == ext:
-                crams[sample] = key
-            elif "crai" == ext:
-                crais[sample] = key
-            else:
-                continue
-        samples = sorted(crams.keys())
-        upload_columns(table, dict(sample=samples,
-                                   cram=[crams[s] for s in samples],
-                                   crai=[crais[s] for s in samples]))
+            sample, ext = parse_cram_crai(filename)
+            if ext in ['cram', 'crai']:
+                matched_files[sample][ext] = filename
+                if ext == 'cram':
+                    allCrams.append(sample)
+                else:
+                    allCrais.append(sample)
+
+        if allCrams != allCrais:
```

Comment:
This still seems like a functional issue where if you have the same number of unmatched crams and crais, this will error.

Reply:
I don't think this is an issue unless I don't understand what you mean. My test case actually is exactly that scenario (one sample has a missing cram and another has a missing crai) and it always catches it. `allCrams != allCrais` isn't equivalent to what I was doing earlier, which was basically `len(allCrams) != len(allCrais)`

Reply:
The comparison actually does an iteration of all elements in order to compare the two lists, but that's more of a nit.  This is fine.